Improve dtype inference and reflection

bmerry opened the pull request:
Also adds some targeted unit tests for it.

bmerry commented:
@jakirkham FYI since I know you'd mentioned you were thinking of doing this.

jakirkham commented:
Also might want to borrow from the testing strategy in PR ( https://github.com/dask/dask/pull/2556 ).

bmerry commented:
> Also might want to borrow from the testing strategy in PR ( #2556 ).

Okay, I've added a test roughly along those lines.

bmerry commented:
By the way, I'm seeing a lot of Travis failures that don't look related to anything I touched. Are those expected?

bmerry commented:
I'm working on reducing this to two tests: one to make sure that `result_type` does what I expect it to do (e.g. given the numpy docs) and one to check that an element-wise operation gives the same dtype as `result_type` (which can be seen more as a test of `elemwise`).

I'll check it in, but it's failing at the moment due to #2574. I raised the question there of what the "correct" dtype is when an input is a 0d dask array, and until that's resolved we probably shouldn't merge this.

bmerry commented:
I've updated the PR title to reflect the new wider scope (namely, fixing #2574 and #2575).

At the moment I've made the explicit dtype casting only kick in when given a 0d dask array (which is the danger area for standard numpy element-wise operations and ufuncs), because I was worried that doing it unconditionally might bloat the graphs significantly.

bmerry commented:
I'm very confused by the Appveyor failure: it's my new test that's failing, but I don't know why it's behaving differently to the Travis build (from what I can see, Travis has a build with the same numpy and Python version as the Appveyor build). If anybody spots something fishy while reviewing, let me know. Unfortunately I don't have a Windows machine to debug on.

mrocklin commented on dask/array/core.py:
```diff
@@ -2750,24 +2751,30 @@ def elemwise(op, *args, **kwargs):
+        # hold potentially very expensive calculations. Instead, we treat
+        # them just like other arrays, and if necessary cast the result of op
+        # to match.
+        vals = [np.ones((1,) * max(1, a.ndim), dtype=a.dtype)
```
what if `np.ones` isn't sensible for `a.dtype`?  Such as for a record dtype?

jcrist replied:
I don't think we need to switch to `ones` at all. The issue was the shape, not the value. But `ones` does work for record dtypes.

```python
In [2]: np.ones(3, dtype=[('a', 'i8'), ('b', 'S2')])
Out[2]:
array([(1, b'1'), (1, b'1'), (1, b'1')],
      dtype=[('a', '<i8'), ('b', 'S2')])
```

bmerry replied:
I switched to ones because I was worried about what would happen if the value was illegal for the operation e.g. 0 for log, or nan generally. But I see `apply_infer_dtype` suppresses warnings, so that's probably not an issue. I'll revert it.

jcrist commented:
> I'm very confused by the Appveyor failure

On windows the default `int` dtype is `int32`. Somewhere in your test you're assuming `int64`.

bmerry commented:
> On windows the default int dtype is int32. Somewhere in your test you're assuming int64.

Ah, of course. Thanks, that's something I can dig into.

bmerry commented:
I've fixed up the Appveyor failure, reverted the change from `np.empty` to `np.ones`, and squashed some commits together. Hopefully this is good to go now.

mrocklin commented:
Sorry for the delay here.  @jcrist is this good to go?

bmerry commented:
I've merged master into the branch. There were a lot of conflicts due to a lot of the functions being moved out of core.py into routines.py. It might be worth someone having a quick look to make sure that my additions have ended up in the appropriate files.

jcrist commented:
Thanks @bmerry! I moved one of the tests back to `da.tests.test_array_core`. Will merge once tests pass.
